transports/webrtc/src/transport.rs

```diff
@@ -0,0 +1,697 @@
+            .create_data_channel(
+                "data",
+                Some(RTCDataChannelInit {
+                    negotiated: None,
+                    id: Some(1),
+                    ordered: None,
+                    max_retransmits: None,
+                    max_packet_life_time: None,
+                    protocol: None,
+                }),
+            )
+            .await?;
+
+        let (data_channel_rx, data_channel_tx) = oneshot::channel::<Arc<DetachedDataChannel>>();
+
+        // Wait until the data channel is opened and detach it.
+        let d = Arc::clone(&data_channel);
+        data_channel
+            .on_open(Box::new(move || {
+                debug!("Data channel '{}'-'{}' open.", d.label(), d.id());
+
+                let d2 = Arc::clone(&d);
+                Box::pin(async move {
+                    match d2.detach().await {
+                        Ok(detached) => {
+                            if let Err(_) = data_channel_rx.send(detached) {
+                                error!("data_channel_tx dropped");
+                            }
+                        },
+                        Err(e) => {
+                            error!("Can't detach data channel: {}", e);
+                        },
+                    };
+                })
+            }))
+            .await;
+
+        let offer = peer_connection
+            .create_offer(None)
+            .map_err(Error::WebRTC)
+            .await?;
+        debug!("OFFER: {:?}", offer.sdp);
+        peer_connection
+            .set_local_description(offer)
+            .map_err(Error::WebRTC)
+            .await?;
+
+        // Set the remote description to the predefined SDP.
+        let fingerprint = match addr.iter().last() {
+            Some(Protocol::XWebRTC(f)) => f,
+            _ => {
+                return Err(Error::InvalidMultiaddr(addr));
+            },
+        };
+        let server_session_description = render_description(
+            sdp::SERVER_SESSION_DESCRIPTION,
+            socket_addr,
+            &fingerprint_to_string(&fingerprint),
+        );
+        debug!("ANSWER: {:?}", server_session_description);
+        let sdp = RTCSessionDescription::answer(server_session_description).unwrap();
+        // Set the local description and start UDP listeners
+        // Note: this will start the gathering of ICE candidates
+        peer_connection
+            .set_remote_description(sdp)
+            .map_err(Error::WebRTC)
+            .await?;
+
+        // wait until data channel is opened and ready to use
+        match tokio_crate::time::timeout(Duration::from_secs(10), data_channel_tx).await {
+            Ok(Ok(dc)) => Ok(Connection::new(peer_connection, dc)),
+            Ok(Err(e)) => Err(Error::InternalError(e.to_string())),
+            Err(_) => Err(Error::InternalError(
+                "data channel opening took longer than 10 seconds (see logs)".into(),
+            )),
+        }
```

Comment:
This makes this code only work on the `tokio` executor. You could use `futures::future::select` and `futures_timer::Delay` to avoid that.

Reply:
I also believe we have a separate `TimeoutTransport` that provides this feature already in a composable way for all transports.

Reply:
> the entire `webrtc` library, which is used here, depends on tokio, so I figured it's logical.

Didn't know that but makes total sense, thanks!